From 8f1f045ba6302f9d802ec39d163d617796d7edda Mon Sep 17 00:00:00 2001
From: Wen Gong <quic_wgong@quicinc.com>
Date: Mon, 24 Oct 2022 03:33:26 -0400
Subject: [PATCH 101/112] wifi: mac80211: change to use MLD address as primary
 link local address

Changes to set MLD address as primary link address
during link setup

Signed-off-by: Wen Gong <quic_wgong@quicinc.com>
Signed-off-by: Amutha Ravikumar <quic_aravikum@quicinc.com>
---
 net/mac80211/mlme.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -6284,9 +6284,14 @@ void ieee80211_mgd_setup_link(struct iee
 	INIT_WORK(&link->u.mgd.chswitch_work, ieee80211_chswitch_work);
 	timer_setup(&link->u.mgd.chswitch_timer, ieee80211_chswitch_timer, 0);
 
-	if (sdata->u.mgd.assoc_data)
+	if (sdata->u.mgd.assoc_data) {
 		ether_addr_copy(link->conf->addr,
 				sdata->u.mgd.assoc_data->link[link_id].addr);
+	} else if (link != &sdata->deflink) {
+		ether_addr_copy(link->conf->addr, sdata->vif.addr);
+	} else if (!is_valid_ether_addr(link->conf->addr)) {
+		eth_random_addr(link->conf->addr);
+	}
 }
 
 /* scan finished notification */
